versioned_docs/version-2021sp/lecture2.md

```diff
@@ -10,3 +10,385 @@ title: Lecture 2
+
+### How to use and interpret package.json
+
+`package.json` contains instructions for necessary packages and scripts that you can use to run your code.
+
+- Dependencies
+  - Packages we use in code
+  - if not installed on local end, this will cause an error
+  - ie. express, ts-node, typescript
+- devDependencies
+  - Toolkits used during development to make development easier
+  - ie. types, autocompletion
+
+## Live Coding Demo
+
+We demoed how to set up a yarn project and create some basic getter HTTP routes.
+
+Set up a yarn project by running `yarn init`. It will ask you a few questions and you can press return to accept all the defaults:
+
+```bash
+yarn init v1.22.5
+question name (lecture2):
+question version (1.0.0):
+question description:
+question entry point (index.ts):
+question repository url:
+question author:
+question license (MIT):
+question private:
+success Saved package.json
+✨  Done in 37.88s.
+```
+
+:::tip
+You can also use `yarn init -y` to set up a project with all the default values!
+:::
+
+This will generate a `package.json` file with all the inputs you provided.
+
+We also want to add several dependencies starting with `express`. Add express with `yarn add express`. Since we are also using TypeScript we want to install the `typescript` and `ts-node` packages as well using `yarn add typescript ts-node`. Finally to get some nice autocompletion features, we want to install `@types/node` and `@types/express` as _devDependencies_ using `yarn add -D @types/node @types/express`.
+
+Your `package.json` should now look like the following:
+
+```json title="package.json"
+{
+  "name": "lecture2",
+  "version": "1.0.0",
+  "main": "index.ts",
+  "license": "MIT",
+  "dependencies": {
+    "express": "^4.17.1",
+    "ts-node": "^9.1.1",
+    "typescript": "^4.2.2"
+  },
+  "devDependencies": {
+    "@types/express": "^4.17.11",
+    "@types/node": "^14.14.31"
+  }
+}
+```
+
+Now we can define some basic express routes in a file `index.ts`:
+
+```ts title="index.ts"
+import * as express from 'express';
+
+const app = express();
+
+app.get('/home', function (req, res) {
+  res.send('Welcome home!');
+});
+
+// example using path parameters
+app.get('/users/:name/:lname', function (req, res) {
+  res.send('Hello ' + req.params.name + ' ' + req.params.lname);
+});
+
+// example using query parameters
+app.get('/users', function (req, res) {
+  res.send('Hello ' + req.query.name + ' ' + req.query.lname);
+});
+
+// tell express to listen for requests on port 8080
+app.listen(8080, function () {
+  console.log('server started');
+});
+```
+
+Add the following to your `package.json`:
+
+```json
+// other package.json properties...
+"scripts": {
+  "start": "ts-node index.ts"
+},
+```
+
+This will allow you to run `ts-node` without having it installed globally (which is bad practice). Now you can run the app with `yarn start`.
```

Comment:
We don't need to include the part about installing globally since that was a problem last sem (fixed this sem)